﻿﻿﻿# Protein Secondary Structure Support Vector Mechine PredictorThis predictor takes protein sequence fasta files, and predicts the amino acid structure in 3 state format.It is trained using cas2.3line dataset._To use this predictor, please run the **/scripts/predictor.py** file._**Features:**-  Use evolutionary information (psiblast and PSSM)-  Use the neighbor amino acids information in prediction (Builing amino acid windows)- A variaty of SVM methods to choose, including linear SVM and svmSVC.- _Cross Validation methods are used to split the dataset._- _Other machine learning method can be used to compare the prediction result, including random forest and simple decision tree.__This predictor is written in Python3_**Download：**``` git clone “https://www.github.com/FuqiXu/BioinfoProject"```**Working directory setting:**! no hard code**Packages required：**  To use this predictor, you need to install biopython 1.70,  pandas 0.22.0, scikiy-learn 0.19.1 package。## How to use:You can use this predictor **directly by running predictor.py**### Parameters to choose : PredictorSVM(trainfilename,testfilename,windowsize, svmfunction = svmSVC,cvfunc=KfoldCV)**trainfilename**: 'trainset.fasta'**testfilename**:  'testset.fasta' or 'additional50Sequences_3lines.txt'**windowsize**: uneven number >3 , default value =13**svmfunction**: svmSVC,linSVC,randomforest,SimpleDeci, default method = svmSVC**cross validation**: KfoldCV, LooCV, basicCV (The code need to be fixed, I splitted the dataset manually)### Results:**file1:** prediction result in 4 lines, sequenceID, sequence,predicted topology , real topology**file2** prediction evaluation, the overall correct rate, the correct rate of "H","C","E" topologies.## Coding files:### Module1：Predict##### 1.  parser_RawData_Numberize.py###### Purpose: 1.  Read raw data2.  Turn sequence data into onehot encode3.  Turn topology data into integers###### Function:1. _f1: Parser()_   _Input:_ raw fata in file format, 'cas2,line3.txt'   _Output:_ Sequence data in an array2. _f2: TopoArray(()_   _Input:_ raw fata in list format   _Output:_ Topology data in an array#####  2. window_size_SeqStructure.py###### Purpose: Create windows###### Function:3.  _f: WindowSeq(windowsize,data)_   _Input:_ windowsize，sequence data in array format   _Output:_ Sequence data in windows， as an array#####  3. parser_OriginalDataset_withWindow###### Purpose: Parse test dataset to be used in svm###### Function:1.  _f1: TestSeqFile()_   _Input:_ windowsize，sequence data in file format   _Output:_ Sequence data in windows as an array1. _f2: TestTopoFile(filename)_   _Input:_ test data in file  format   _Output:_ Topology data in an array#####  4. predictionFunction_SVM_RF_SDT.py###### Purpose: Use machine learning method to do predictionThese functions is used as a parameter in the predictor file.###### Function:1.  _f1: svmSVC()_2. _f2: linSVC()_3. _f3: randomforest()_4. _SimpleDeci()__Input:_ trainSeq,trainTopo,testseq_Output:_ a svm model to be used in prediction##### 5. predictionEval.py###### Purpose: Evaluate predictor performance###### Function:1.  _f1: score (testTopoReal,predictTopology))_   _Input:_ two topology arrays   _Output:_  file 'prediction accuracy.txt'.### Module2：Parse file to be predicted#####  1.Parser_AdditionalDataset_to3lines.py###### Purpose: Parse test dataset to be used in svm###### Function:1.  _f1: sequenceConverter(filename)_   _Input:_ Raw data downloaded from swissprot   _Output:_ A 3-line format file, using 3 state topology representation### **Module3 :** Do psiblast and creat PSSM##### 1. formatdb.sh###### Purpose: Format protein database to be used in psiblast##### 2. BlastParser_RawTestDataset_Splitlines.py###### Purpose: Split the original dataset into 3 lines files, each sequence is a file###### Function:1.  _f1: blastParserRaw(filename)_   _Input:_ raw data in file format ('cas2.3line.txt')   _Output:_ multiple 3-line sequence files##### 3. psiblast.sh###### Purpose: Do psiblast on each sequence file and return pssm files. _Input:_  directory to be psiblasted _Output:_ PSSM files##### 3. parser_PSIBlastResult_MultipleFiles.py###### Purpose: Extract pssm matrix from psiblast raw output on multiple files.###### Function:1.  _f1: UniPSSMparser(directory)_   _Input:_ directory which stores the psiblast pssm result   _Output:_  files containing PSSM concise matrix.## Dataset### 1. additional50Sequences.txtExtra 50 sequences, obtained from another protein secondary structure predictor### 2. additional50Sequences_3lines.txtParsed extra 50 sequences data file, can be used as input.### 3. cas2.3line.txtOriginal dataset### 4. trainset.fasta### 5. testset.fasta### 6. psiblast_parsedPSSM matrixs### 7. swissprot.fastaSwissprot dataset 